File: cpp/WordSearch.cc
namespace WordSearch {
  class Solution {
  public:
      bool dfsSearch(vector<vector<char> > &board, string word, vector<vector<bool> >& visited, int i, int j, int w, int h, int curIdx) {
  
          
          if (word[curIdx] != board[i][j]) {
              return false;
          }
          
          if (curIdx == word.size() - 1) {
              // end of string match
              return true;
          }
          
          visited[i][j] = true;
          bool res;
          
          if (i > 0 && visited[i - 1][j] == false) {
              if (dfsSearch(board, word, visited, i - 1, j, w, h, curIdx + 1)) {return true;}
          }
          
          if (i < w - 1 && visited[i + 1][j] == false) {
              if (dfsSearch(board, word, visited, i + 1, j, w, h, curIdx + 1)) {return true;}
          }
          
          if (j > 0 && visited[i][j - 1] == false) {
              if (dfsSearch(board, word, visited, i, j - 1, w, h, curIdx + 1)) {return true;}
          }
          
          if (j < h - 1 && visited[i][j + 1] == false) {
              if (dfsSearch(board, word, visited, i, j + 1, w, h, curIdx + 1)) {return true;}
          }
          
          visited[i][j] = false;
          
          return false;
      }
      
      bool exist(vector<vector<char> > &board, string word) {
         if (board.empty() || board[0].empty()) {
             return false;
         }
         
         if (word.empty()) {
             return false;
         }
         
         int w = board.size();
         int h = board[0].size();
         
         vector<vector<bool> > visited(w, vector<bool>(h, false));
         
         for (int i = 0; i < board.size(); i++) {
             for (int j = 0; j < board[i].size(); j++) {
                 bool res = dfsSearch(board, word, visited, i, j, w, h, 0);
                 if (res == true) {
                     return true;
                 }
             }
         }
         
         return false;
      }
  };
}

File: java/src/main/java/org/speedfirst/leetcode/array/WordSearch.java
package org.speedfirst.leetcode.array;

/**
 * @author jiankuan
 * @date 12/13/14
 */
public class WordSearch {

    public boolean searchWord(char[][] board, boolean[][] visited, int w, int h, int curX, int curY, int curWordIdx, String word) {
        char c = word.charAt(curWordIdx);
        if (board[curX][curY] != c) {
            return false;
        }

        if (curWordIdx == word.length() - 1) {
            return true;
        }

        visited[curX][curY] = true;
        if (curX > 0 && !visited[curX - 1][curY] && searchWord(board, visited, w, h, curX - 1, curY, curWordIdx + 1, word)) return true;
        if (curX < h - 1 && !visited[curX + 1][curY] && searchWord(board, visited, w, h, curX + 1, curY, curWordIdx + 1, word)) return true;
        if (curY > 0 && !visited[curX][curY - 1] && searchWord(board, visited, w, h, curX, curY - 1, curWordIdx + 1, word)) return true;
        if (curY < w - 1 && !visited[curX][curY + 1] &&  searchWord(board, visited, w, h, curX, curY + 1, curWordIdx + 1, word)) return true;
        visited[curX][curY] = false;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        int h = board.length;
        int w = board[0].length;

        boolean[][] visited = new boolean[h][w];
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                visited[i][j] = false;
            }
        }

        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (searchWord(board, visited, w, h, i, j, 0, word)) {
                    return true;
                }
            }
        }

        return false;
    }

    public static void main(String[] args) {
        char[][] board = new char[1][1];
        board[0][0] = 'a';
        WordSearch ws = new WordSearch();
        boolean res = ws.exist(board, "a");
        System.out.println(res);
    }
}
